File: 162. Find Peak Element.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findPeakElement = function(nums) {
    if ( nums == null || nums.length == 0 ) {
        return -1;
    }

    var mid;
    var start = 0;
    var end = nums.length - 1;

    while ( start + 1 < end ) {
        mid = start + Math.round( (end - start) / 2 );

        if( nums[mid-1] > nums[mid] ) {
            end = mid;
        } else if ( nums[mid] < nums[mid+1] ) {
            start = mid;
        } else {
            start = mid;
        }
    }

    if( nums[start] < nums[end] ) {
        return end;
    } else {
        return start;
    }
};

File: 162. Find Peak Element.py
"""
只要找到其中一个高峰，不需要是最高的
"""



class Solution(object):
    def findPeakElement(self, nums):
        if not nums or len(nums) == 0:
            return -1

        low = 0
        high = len(nums) - 1

        while low + 1 < high:
            mid = (low + high) / 2

            if nums[mid] < nums[mid + 1]:
                low = mid
            else:
                high = mid

        if nums[low] < nums[high]:
            return high
        else:
            return low
